### PR TITLE
test: add e2e for akeyless csi provider

### PR DESCRIPTION
**Pull request by d-goro**
**What this PR does / why we need it**:
Adds tests using new CSI provider

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Review by aramase — COMMENTED** [review]
Waiting for https://github.com/kubernetes/test-infra/pull/25707 to be merged before I can run `ok-to-test` on this PR.

Next steps after the test-infra PR is merged:
1. We can trigger this PR to run and validate the changes. As there is no prow job configured for akeyless provider yet, you can modify any of the [make target](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/Makefile#L468-L486) to run your providers tests to validate the tests in this PR.
2. After this PR is green, we can add an optional job similar to https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/secrets-store-csi-driver/secrets-store-csi-driver-config.yaml#L315-L352 for the akeyless provider. This job will be `optional: true` and `required: false`.

**Comment by aramase** [comment]
/retitle: test: add e2e for akeyless csi provider

File: docs/book/src/concepts.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ Currently supported providers:
 - [Azure Provider](https://azure.github.io/secrets-store-csi-driver-provider-azure/)
 - [GCP Provider](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp)
 - [Vault Provider](https://github.com/hashicorp/secrets-store-csi-driver-provider-vault)
+- [Akeyless Provider](https://github.com/akeylesslabs/akeyless-csi-provider)
```
Comment (aramase) [inline docs/book/src/concepts.md:33]: Let's make the doc updates in a separate PR after:

1. Tests are passing and merged
2. Code audit of the provider is complete

File: test/bats/akeyless.bats [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+   # On Windows, the failed unmount calls from: https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/545
+   # do not prevent the pod from being deleted. Search through the driver logs
+   # for the error.
+   run bash -c "kubectl logs -l app=secrets-store-csi-driver --tail -1 -c secrets-store -n kube-system | grep '^E.*failed to clean and unmount target path.*$'"
+   assert_failure
+ }
+
+ teardown_file() {
+   archive_info || true
+ }
```
Comment (aramase) [inline test/bats/akeyless.bats:82]: nit: add new line

File: test/bats/tests/akeyless/pod-inline-volume-secretproviderclass.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    volumeMounts:
+    - name: secrets-store-inline
+      mountPath: "/mnt/secrets-store"
+      readOnly: true
+    env:
+      - name: AKEYLESS_ACCESS_KEY
+        valueFrom:
+          secretKeyRef:
+            name: credentials
+            key: akeyless-test-cred
```
Comment (aramase) [inline test/bats/tests/akeyless/pod-inline-volume-secretproviderclass.yaml:22]: This credential is not required by the workload pod. The CSI provider is the one that users the credentials to get the secret. 

I've opened this PR to add an optional job for akeyless provider: https://github.com/kubernetes/test-infra/pull/25751. The required credentials will be injected into the test pod and value should be available in env var `AKEYLESS_ACCESS_KEY`. You can use that value to create a Kubernetes secret and reference that Kubernetes with `nodePublishSecretRef`.

https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/3d93b4a03a68fd0de22fc09711734db0d2f51f70/test/bats/azure.bats#L62-L69

That secret would be referenced in the volume mount like this:
https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/3d93b4a03a68fd0de22fc09711734db0d2f51f70/test/bats/tests/azure/pod-secrets-store-inline-volume-crd.yaml#L25-L26

The driver will pass the value of that secret in https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/3d93b4a03a68fd0de22fc09711734db0d2f51f70/provider/v1alpha1/service.proto#L48-L49

**Comment by aramase** [comment]
> We can trigger this PR to run and validate the changes. As there is no prow job configured for akeyless provider yet, you can modify any of the [make target](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/Makefile#L468-L486) to run your providers tests to validate the tests in this PR.

@d-goro We already merged https://github.com/kubernetes/test-infra/pull/25751 to add an optional job for akeyless provider. The tests can be triggered by running `/test pull-secrets-store-csi-driver-e2e-akeyless`. Please see the PR description for the details. 

Could you take a look at the comments on the PR and resolve those?

File: Makefile [hunk 4]
```diff
@@ -477,6 +477,10 @@ e2e-azure: $(AZURE_CLI)
 e2e-vault:
 	bats -t test/bats/vault.bats
 
+.PHONY: e2e-akeyless
+ e2e-akeyless:
+ 	bats -t test/bats/akeyless.bats
+
```
Comment (aramase) [inline Makefile:483]:
```suggestion
.PHONY: e2e-akeyless
e2e-akeyless:
	bats -t test/bats/akeyless.bats
```

All the tests for PR are failing because of line 483 and extra tab in line 483

File: docs/book/src/concepts.md (hunk 1 above)
Comment (d-goro) [reply docs/book/src/concepts.md]: removed

File: test/bats/akeyless.bats (hunk 2 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: fixed

File: test/bats/tests/akeyless/pod-inline-volume-secretproviderclass.yaml (hunk 3 above)
Comment (d-goro) [reply test/bats/tests/akeyless/pod-inline-volume-secretproviderclass.yaml]: removed

Comment (d-goro) [reply test/bats/tests/akeyless/pod-inline-volume-secretproviderclass.yaml]: added support for k8s secret

**Comment by d-goro** [comment]
> > We can trigger this PR to run and validate the changes. As there is no prow job configured for akeyless provider yet, you can modify any of the [make target](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/Makefile#L468-L486) to run your providers tests to validate the tests in this PR.
> 
> @d-goro We already merged [kubernetes/test-infra#25751](https://github.com/kubernetes/test-infra/pull/25751) to add an optional job for akeyless provider. The tests can be triggered by running `/test pull-secrets-store-csi-driver-e2e-akeyless`. Please see the PR description for the details.
> 
> Could you take a look at the comments on the PR and resolve those?

All resolved, please, take a look

File: Makefile [hunk 5]
```diff
@@ -473,9 +473,15 @@ e2e-provider:
 e2e-azure: $(AZURE_CLI)
 	bats -t test/bats/azure.bats
 
+# .PHONY: e2e-vault
+# e2e-vault:
+# 	bats -t test/bats/vault.bats
+
+# .PHONY: e2e-akeyless
+# e2e-akeyless:
```
[the remark this answers is not in the document]
Comment (d-goro) [reply Makefile]: fixed

File: test/bats/akeyless.bats [hunk 6]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env bats
+
+ load helpers
+
+ BATS_TESTS_DIR=test/bats/tests/akeyless
+ WAIT_TIME=120
+ SLEEP_TIME=1
+
+
```
Comment (aramase) [inline test/bats/akeyless.bats:9]: nit: remove extra newline

File: test/bats/akeyless.bats [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ export LABEL_VALUE=${LABEL_VALUE:-"test"}
+ export ANNOTATION_VALUE=${ANNOTATION_VALUE:-"app=test"}
```
Comment (aramase) [inline test/bats/akeyless.bats:11]: These values are not used anywhere in this test suite. Could you remove them?

File: test/bats/akeyless.bats [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+@test "install akeyless csi provider" {
+   # create the ns akeyless
+   kubectl create ns akeyless
+   # install the akeyless provider using the helm charts
+
+   helm repo add akeyless https://akeylesslabs.github.io/helm-charts
+   helm repo update
+   helm install akeyless akeyless/akeyless-csi-provider --namespace=akeyless --set csi.image.pullPolicy=Always
```
Comment (aramase) [inline test/bats/akeyless.bats:20]: Lets use `--create-namespace` flag here with helm install instead of creating namespace [here](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/910/files#diff-882da5c47eff475fd29efc19d199866354b582b7e713fce8586fccb8911d5f05R15).

File: test/bats/akeyless.bats [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+ WAIT_TIME=120
+ SLEEP_TIME=1
+
+
+ export LABEL_VALUE=${LABEL_VALUE:-"test"}
+ export ANNOTATION_VALUE=${ANNOTATION_VALUE:-"app=test"}
+
+@test "install akeyless csi provider" {
+   # create the ns akeyless
+   kubectl create ns akeyless
```
Comment (aramase) [inline test/bats/akeyless.bats:15]: This can be removed after using `--create-namespace` with helm install command

File: test/bats/akeyless.bats (hunk 8 above)
Comment (aramase) [reply test/bats/akeyless.bats:20]: Why are we setting image pull policy `Always`?

File: test/bats/akeyless.bats [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+   # install the akeyless provider using the helm charts
+
+   helm repo add akeyless https://akeylesslabs.github.io/helm-charts
+   helm repo update
+   helm install akeyless akeyless/akeyless-csi-provider --namespace=akeyless --set csi.image.pullPolicy=Always
+
+   # wait for akeyless and akeyless-csi-provider pods to be running
```
Comment (aramase) [inline test/bats/akeyless.bats:22]: What are the components that helm chart installs? Is there a akeyless instance and provider?

File: test/bats/akeyless.bats [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+   kubectl wait --for=condition=Ready --timeout=${WAIT_TIME}s pods --all -n akeyless
+ }
+
+@test "create akeyless k8s secret" {
+  run kubectl create secret generic secrets-store-creds --from-literal akeylessAccessKey=${AKEYLESS_ACCESS_KEY}
```
Comment (aramase) [inline test/bats/akeyless.bats:27]: Maybe check if `AKEYLESS_ACCESS_KEY` is defined similar to https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/2beb3ecc8d500daac4f8574e0c7b7867e1519583/test/bats/azure.bats#L43-L48 so the tests can fail early when required creds aren't available?

File: test/bats/akeyless.bats [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+@test "secretproviderclasses crd is established" {
+   kubectl wait --for condition=established --timeout=${WAIT_TIME}s crd/secretproviderclasses.secrets-store.csi.x-k8s.io
+
+   run kubectl get crd/secretproviderclasses.secrets-store.csi.x-k8s.io
+   assert_success
+}
+
+@test "deploy akeyless secretproviderclass crd" {
+   kubectl apply -f $BATS_TESTS_DIR/secretproviderclass.yaml
+   kubectl wait --for condition=established --timeout=${WAIT_TIME}s crd/secretproviderclasses.secrets-store.csi.x-k8s.io
```
Comment (aramase) [inline test/bats/akeyless.bats:44]: This is already checked [here](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/910/files#diff-882da5c47eff475fd29efc19d199866354b582b7e713fce8586fccb8911d5f05R36), so we can remove this line

File: test/bats/akeyless.bats [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+   # On Windows, the failed unmount calls from: https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/545
+   # do not prevent the pod from being deleted. Search through the driver logs
+   # for the error.
+   run bash -c "kubectl logs -l app=secrets-store-csi-driver --tail -1 -c secrets-store -n kube-system | grep '^E.*failed to clean and unmount target path.*$'"
+   assert_failure
+}
+
+teardown_file() {
+   archive_info || true
```
Comment (aramase) [inline test/bats/akeyless.bats:90]: Let's also get the provider log for akeyless provider similar to https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/2beb3ecc8d500daac4f8574e0c7b7867e1519583/test/bats/azure.bats#L396

File: test/bats/akeyless.bats [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  run kubectl create secret generic secrets-store-creds --from-literal akeylessAccessKey=${AKEYLESS_ACCESS_KEY}
+  assert_success
+
+  # label the node publish secret ref secret
+  run kubectl label secret secrets-store-creds secrets-store.csi.k8s.io/used=true
+  assert_success
+}
+
+@test "secretproviderclasses crd is established" {
+   kubectl wait --for condition=established --timeout=${WAIT_TIME}s crd/secretproviderclasses.secrets-store.csi.x-k8s.io
```
Comment (aramase) [inline test/bats/akeyless.bats:36]:
```suggestion
  kubectl wait --for condition=established --timeout=${WAIT_TIME}s crd/secretproviderclasses.secrets-store.csi.x-k8s.io
```



The indentation inside `test <description> {}` in the bats file isn't consistent. Could you update it to be 2 spaces across the entire file?

**Review by aramase — COMMENTED** [review]
Thanks for the PR! Added few comments.

File: test/bats/akeyless.bats (hunk 14 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: fixed

File: test/bats/akeyless.bats (hunk 13 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: Why? What does it do? Where to get it? 
Vault tests don't use it. Actually, only Azure and GCP doing this.

File: test/bats/akeyless.bats (hunk 12 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: fixed

File: test/bats/akeyless.bats (hunk 11 above)
Comment (d-goro) [reply test/bats/akeyless.bats:27]: fixed

File: test/bats/akeyless.bats (hunk 10 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: Akeyless CSI provider. We don't have "instance" like Vault, since we are cloud secret manager

File: test/bats/akeyless.bats (hunk 9 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: fixed

File: test/bats/akeyless.bats (hunk 8 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: fixed

File: test/bats/akeyless.bats (hunk 7 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: fixed

File: test/bats/akeyless.bats (hunk 6 above)
Comment (d-goro) [reply test/bats/akeyless.bats]: fixed

File: test/bats/akeyless.bats (hunk 13 above)
Comment (aramase) [reply test/bats/akeyless.bats]: `archive_provider` is to collect the provider logs as part of CI run https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/2beb3ecc8d500daac4f8574e0c7b7867e1519583/test/bats/helpers.bash#L84-L97.

- If there are CI failures, logs will help with debugging
- If it's not in vault provider, we can add it in a follow up PR. This is nothing specific to any provider

Comment (d-goro) [reply test/bats/akeyless.bats]: Ok, what should we do? What does that line `archive_provider "app=csi-secrets-store-provider-azure" || true ` means?

**Review by aramase — COMMENTED** [review]
@d-goro changes look good! Could you squash the commits and rebase?

**Comment by d-goro** [comment]
> @d-goro changes look good! Could you squash the commits and rebase?

done

**Review by aramase — COMMENTED** [review]
Thanks for the PR! 

/approve
